### PR TITLE
Diagnostic connection counters

### PR DESCRIPTION
**Pull request by peterbryntesson**
This PR adds three diagnostic counters per endpoint:
* Is the endpoint connection up and running?
* How many nodes have successfully connected and are receiving data?
* How many nodes have misconfiguration causing them to not connect?

Example of diagnostic output can be seen here:
image

### CODE REVIEW

**Review by simonjaeger — DISMISSED** [review]
LGTM

**Review by cristipogacean — DISMISSED** [review]
nice improvement

**Comment by karok2m** [comment]
@peterbryntesson, can you please add information about new diagnostics counters in `docs/manual/readme.md` ? There is a `IIoT Metrics` section about available metrics.

**Comment by peterbryntesson** [comment]
> @peterbryntesson, can you please add information about new diagnostics counters in `docs/manual/readme.md` ? There is a `IIoT Metrics` section about available metrics.

Done

**Comment by peterbryntesson** [comment]
Sorry, mistakenly closed the PR, but has reopened it.

**Review by cristipogacean — DISMISSED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 1]
```diff
@@ -30,6 +30,18 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
         public int NumberOfConnectionRetries => _subscriptions?.FirstOrDefault()?.
             Subscription?.NumberOfConnectionRetries ?? 0;
 
+        /// <inheritdoc/>
+        public bool IsConnectionOk => _subscriptions?.FirstOrDefault()?.
+            Subscription?.IsConnectionOk ?? false;
+
+        /// <inheritdoc/>
+        public int NumberOfGoodNodes => _subscriptions?.FirstOrDefault()?.
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: @peterbryntesson, why are we getting `NumberOfGoodNodes` only from the first subscription and not summing over all of them ? I have the same question about all new counters in this class, as well as `NumberOfConnectionRetries` which I see was already present. Can we please test that if `WriterGroupMessageTrigger` has several subscriptions in it then the reported values would still be valid ?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs [hunk 2]
```diff
@@ -64,6 +63,51 @@ public class DefaultSessionManager : ISessionManager, IDisposable {
...
+                }
+                return wrapper.State == SessionState.Running;
+            }
+            finally {
+                _lock.Release();
+            }
+        }
+
+        /// <inheritdoc/>
+        public int GetNumberOfGoodNodes(ConnectionModel connection) {
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs]: @peterbryntesson, `GetNumberOfGoodNodes()` method is not used anywhere. Was it intended to be used in `SubscriptionWrapper` similar to `DefaultSessionManager.IsConnectionOk` ? Can we delete it if it is not referenced anywhere ?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs [hunk 3]
```diff
@@ -64,6 +63,51 @@ public class DefaultSessionManager : ISessionManager, IDisposable {
...
+                }
+                return wrapper._subscriptions.Sum(x => x.Value.NumberOfGoodNodes);
+            }
+            finally {
+                _lock.Release();
+            }
+        }
+
+        /// <inheritdoc/>
+        public int GetNumberOfBadNodes(ConnectionModel connection) {
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs]: @peterbryntesson, similar to the one above, `GetNumberOfBadNodes` method is not used anywhere. Was it intended to be used in `SubscriptionWrapper` similar to `DefaultSessionManager.IsConnectionOk` ? Can we delete it if it is not referenced anywhere ?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 1 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: I just followed the pattern with NumberOfConnectionRetries. Let me change that. I will change it for NumberOfConnectionRetries as well.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs]: Good spotting, Forgot to remove those

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs (hunk 3 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs]: Good spotting, Forgot to remove those

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 1 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: Done

**Review by karok2m — APPROVED** [review]

**Review by cristipogacean — APPROVED** [review]

**Pull request merged**